opflexagent/opflex_conn_track.py

```diff
@@ -22,16 +24,82 @@
-    return ret
+class SnatConntrackLogger(object):
+    """Opflex SNAT Connection Traking Logger
+
+    This class dumps the output of connection tracking events
+    for a Linux network namespace used to perform SNAT on hosts
+    running the neutron-opflex-agent. This class is created in
+    order to spawn two processes per namespace: one to run a
+    connection tracker in the Linux network namespace, and the
+    other to collect the standard output and dump it into syslog,
+    using the log level and identity set in a configuration file.
+    The process for this is managed using supervisord, which in
+    version 4.0 or later supports redirecting stdout and stderr
+    to syslog. However, older Linux distributions ship with
+    supervisord versions older than this, so the logger tool
+    must be used for this redirection. While supervisord can
+    handle the lifecycle of the top level process, this class
+    neededs manage the lifecycle of the child prcoesses that
+    are spawned (i.e. conntrack and logger), ensuring that
+    the parent exit code is set correctly so that supervisord
+    can handle any respawning, if needed.
+
+    REVISIT: Should have some sort of pooling to avoid exhaustion
+    """
+    def __init__(self, conntrack_cmd, logger_cmd):
+        self.ret = None
+        self.p1 = self.p2 = None
+        try:
+            # Conntrack process redirects stderr to stdout,
+            # retured Popen has file object for stdout
+            self.p1 = subprocess.Popen(
+                conntrack_cmd, stdout=subprocess.PIPE,
+                stderr=subprocess.STDOUT, close_fds=True, shell=True)
+            # Logger uses stdout member from conntrack process for its stdin
+            self.p2 = subprocess.Popen(
+                logger_cmd, stdin=self.p1.stdout, close_fds=True, shell=True)
+        except Exception as e:
+            LOG.error("In running commands: %s: %s" %
+                      (conntrack_cmd + logger_cmd, str(e)))
+
+    def _kill_child_procs(self, save_rc=True):
+        if save_rc:
+            # if poll returns None, it means the process hasn't exited yet
+            # In this case, we're going to kill it, so don't save the RC
+            if self.p1 and self.p1.poll() is not None:
+                # It must have died, so set it to nonzero RC or SIGTERM
+                self.ret = self.p1.returncode or -signal.SIGTERM
+            if not self.ret and self.p2 and self.p2.poll() is not None:
+                # It must have died, so set it to nonzero RC or SIGTERM
+                self.ret = self.p2.returncode or -signal.SIGTERM
+        for proc in [self.p1, self.p2]:
+            LOG.debug("conn_track: proc: %s" % proc.returncode)
+            try:
+                proc.terminate()
+            except Exception:
+                pass
+
+    def terminate(self, signum, frame):
+        if signum is not None:
+            self._kill_child_procs(save_rc=False)
```

Comment:
not sure why we don't care about the return code if this program is getting killed by SIGINT and SIGTERM.....

Reply:
Those are valid values - we need to have cases where a program is terminated validly. That's taken care of by supervisord.

Reply:
It seems to me that even for this case, its still no harm to save the return code right? If so, we can really consolidate the code paths and make it more straightforward.

Reply:
I'm changing this to always exit with signum.